### PR TITLE
PLAT-84674: Fix Dropdown performance with long lists

### PR DESCRIPTION
**Pull request by ryanjduffy**
### Checklist

* [X] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [X] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [X] Documentation was added or is not needed

* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Render performance of `Dropdown` is poor when using many options.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
Adopt `VirtualList` within `Dropdown` instead of `Group` + `Scroller` to reduce the render load.

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)
* Because `VirtualList` requires `itemSize` to lay out the list, I had to hard code the height of `Item` and include `skinVariants` from `Skinnable` to scale for large text mode.
* Dropdown.js was getting long so I refactored the list component into a new file and converted the previous list component into a HOC so `Skinnable` could be moved to the decorator for consistency with other framework components.

### CODE REVIEW

File: packages/moonstone/Dropdown/DropdownList.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+		/*
+		* Callback function that will receive the scroller's scrollTo() method
+		*
+		* @type {Function}
+		*/
+		scrollTo: PropTypes.func,
+
+		/*
+		* Index of the selected item.
```
Comment (webOS101) [inline packages/moonstone/Dropdown/DropdownList.js]: weird indent!

File: packages/moonstone/Dropdown/DropdownList.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+				key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired
+			}))
+		]),
+
+		/*
+		* Called when an item is selected.
+		*
+		* @type {Function}
+		*/
+		onSelect: PropTypes.func,
```
Comment (webOS101) [inline packages/moonstone/Dropdown/DropdownList.js]: Required?  We don't validate presence below.

File: packages/moonstone/Dropdown/DropdownList.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		}
+
+		getScrollTo = (scrollTo) => {
+			this.scrollTo = scrollTo;
+		}
+
+		scrollIntoView = () => {
+			if (!this.isScrolledIntoView) {
+				if (isSelectedValid(this.props)) {
+					const {nodeToFocus, nodeToScroll} = this.getNodeToFocus(this.props.selected);
```
Comment (webOS101) [inline packages/moonstone/Dropdown/DropdownList.js]: The selected is not scrolling into view when I test this.

File: packages/moonstone/Dropdown/DropdownList.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	handlers: {
+		itemRenderer: ({index, ...rest}, {children, onSelect}) => {
+			let child = children[index];
+			if (typeof child === 'string') {
+				child = {children: child};
+			}
+
+			return (
+				<Item
+					{...rest}
```
Comment (webOS101) [inline packages/moonstone/Dropdown/DropdownList.js]: The selected class does not get applied to the selected item.

File: packages/moonstone/Dropdown/DropdownList.js (hunk 2 above)
Comment (ryanjduffy) [reply packages/moonstone/Dropdown/DropdownList.js]: I'll add a guard.

**Comment by ryanjduffy** [comment]
@webOS101 - I believe this is fixed up and ready for another review. Thanks for the feedback!

File: packages/moonstone/Dropdown/DropdownList.js [hunk 5]
```diff
@@ -145,26 +149,20 @@ const DropdownListSpotlightDecorator = hoc((config, Wrapped) => {
...
 		}
 
 		scrollIntoView = () => {
+			const {selected} = this.props;
+
 			if (!this.isScrolledIntoView) {
 				if (isSelectedValid(this.props)) {
-					const {nodeToFocus, nodeToScroll} = this.getNodeToFocus(this.props.selected);
+					const scrollIndex = selected > scrollOffset ? selected - scrollOffset : 0;
+					const selectedNode = this.node.querySelector(`[data-index='${selected}']`);
```
Comment (webOS101) [inline packages/moonstone/Dropdown/DropdownList.js]: Is this able to find one that's not currently in the DOM?  Should we make this a callback on the scroll, maybe?

Comment (ryanjduffy) [reply packages/moonstone/Dropdown/DropdownList.js]: good question. I'll give it a try with a bigger data set and see.

Comment (ryanjduffy) [reply packages/moonstone/Dropdown/DropdownList.js]: Fixed up (I hope!)

Comment (ryanjduffy) [reply packages/moonstone/Dropdown/DropdownList.js]: Fixed up (I hope!)

**Pull request merged**